logs_ingest/dynatrace_client.py

```diff
@@ -27,47 +29,52 @@
 from . import logging
 
 should_verify_ssl_certificate = os.environ.get("REQUIRE_VALID_CERTIFICATE", "True") in ["True", "true"]
+number_of_concurrent_send_calls = get_int_environment_value("NUMBER_OF_CONCURRENT_SEND_CALLS", 2)
 ssl_context = ssl.create_default_context()
 if not should_verify_ssl_certificate:
     ssl_context.check_hostname = False
     ssl_context.verify_mode = ssl.CERT_NONE
 
 
-def send_logs(dynatrace_url: str, dynatrace_token: str, logs: List[Dict], self_monitoring: SelfMonitoring):
-    # pylint: disable=R0912
+async def send_logs(dynatrace_url: str, dynatrace_token: str, logs: List[Dict], self_monitoring: SelfMonitoring):
     start_time = time.perf_counter()
     log_ingest_url = urlparse(dynatrace_url.rstrip("/") + "/api/v2/logs/ingest").geturl()
     batches = prepare_serialized_batches(logs)
 
-    number_of_http_errors = 0
-    for batch in batches:
-        batch_logs = batch[0]
-        number_of_logs_in_batch = batch[1]
-        encoded_body_bytes = batch_logs.encode("UTF-8")
-        display_payload_size = round((len(encoded_body_bytes) / 1024), 3)
-        logging.info(f'Log ingest payload size: {display_payload_size} kB')
-        sent = False
-        try:
-            sent = _send_logs(dynatrace_token, encoded_body_bytes, log_ingest_url, self_monitoring, sent)
-        except HTTPError as e:
-            raise e
-        except Exception as e:
-            logging.exception("Failed to ingest logs", "ingesting-logs-exception")
-            self_monitoring.dynatrace_connectivities.append(DynatraceConnectivity.Other)
-            number_of_http_errors += 1
-            # all http requests failed and this is the last batch, raise this exception to trigger retry
-            if number_of_http_errors == len(batches):
-                raise e
-        finally:
-            self_monitoring.sending_time = time.perf_counter() - start_time
-            if sent:
-                self_monitoring.log_ingest_payload_size += display_payload_size
-                self_monitoring.sent_log_entries += number_of_logs_in_batch
-
-
-def _send_logs(dynatrace_token, encoded_body_bytes, log_ingest_url, self_monitoring, sent):
+    semaphore = asyncio.Semaphore(number_of_concurrent_send_calls)
+    async with aiohttp.ClientSession() as session:  # Create the session once
+        number_of_http_errors = 0
+        async def process_batch(batch_logs: str, number_of_logs_in_batch: int ):
+            async with semaphore:
+                encoded_body_bytes = batch_logs.encode("UTF-8")
+                display_payload_size = round((len(encoded_body_bytes) / 1024), 3)
+                logging.info(f'Log ingest payload size: {display_payload_size} kB')
+                sent = False
+                try:
+                    sent = await _send_logs(session, dynatrace_token, encoded_body_bytes, log_ingest_url,
+                                            self_monitoring, sent)
```

Comment:
Why is this sent (=False) being passed?

Reply:
I looked at _send_logs, it does return True which is ok, it does not look like it would ever return False though, not sure if return True is helpful in such case :D

Reply:
And definitely assigning sent=True inside that function is not the correct way, can you please clean it up?